File: s/split_script.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(optparse)
  library(dplyr)
  library(readr)
  library(stringr)
  library(magick)
  library(progress)
  library(tools)
  library(ggplot2)
  library(grid)
})

# ---------------------------------------------------------
# Helper: Parse GEOMETRY_TEXT to Bounding Box
# ---------------------------------------------------------
parse_geometry <- function(geom_text) {
  if (is.na(geom_text) || geom_text == "") {
    return(c(NA, NA, NA, NA))
  }
  
  nums <- as.numeric(unlist(str_extract_all(geom_text, "[-+]?\\d*\\.\\d+|\\d+")))
  
  if (length(nums) >= 4) {
    x1 <- nums[1]; y1 <- nums[2]; x2 <- nums[3]; y2 <- nums[4]
    return(c(min(x1, x2), min(y1, y2), max(x1, x2), max(y1, y2)))
  }
  return(c(NA, NA, NA, NA))
}

# ---------------------------------------------------------
# Helper: Crop and Save Image
# ---------------------------------------------------------
crop_and_save <- function(src_path, out_path, side) {
  im <- image_read(src_path)
  info <- image_info(im)
  w <- info$width
  h <- info$height
  mid <- w / 2
  
  if (side == "left") {
    geo_string <- sprintf("%dx%d+%d+%d", floor(mid), h, 0, 0)
    offset <- 0
  } else { # right
    geo_string <- sprintf("%dx%d+%d+%d", ceiling(w - mid), h, floor(mid), 0)
    offset <- mid
  }
  
  cropped_im <- image_crop(im, geo_string)
  image_write(cropped_im, path = out_path)
  
  return(list(mid = mid, offset = offset, h = h))
}

# ---------------------------------------------------------
# Helper: Visualize a Subset of Annotations
# ---------------------------------------------------------
visualize_annotations <- function(df, img_dir, out_pdf, n = 5) {
  cat(sprintf("\n🎨 Generating validation visualizations for %d images...\n", n))
  
  # Get unique images that actually have annotations
  valid_images <- unique(df$image)
  if (length(valid_images) == 0) {
    cat("⚠️ No annotations available to visualize.\n")
    return()
  }
  
  sample_imgs <- sample(valid_images, min(n, length(valid_images)))
  
  pdf(out_pdf, width = 10, height = 8)
  
  for (img_name in sample_imgs) {
    img_path <- file.path(img_dir, img_name)
    if (!file.exists(img_path)) next
    
    # Read image for plotting
    im <- image_read(img_path)
    info <- image_info(im)
    w <- info$width
    h <- info$height
    img_grob <- rasterGrob(as.raster(im), interpolate = FALSE)
    
    # Subset annotations for this image
    sub_df <- df %>% filter(image == img_name)
    
    # Plot using ggplot2 with reversed Y-axis (top-left origin)
    p <- ggplot() +
      annotation_custom(img_grob, xmin = 0, xmax = w, ymin = h, ymax = 0) +
      geom_rect(
        data = sub_df,
        aes(xmin = TLx, ymin = TLy, xmax = BRx, ymax = BRy),
        color = "yellow", fill = NA, linewidth = 0.8
      ) +
      coord_fixed(xlim = c(0, w), ylim = c(h, 0), expand = FALSE) +
      labs(
        title = img_name,
        subtitle = sprintf("Validation Plot | %d Annotations | Box Color: Yellow", nrow(sub_df))
      ) +
      theme_void()
    
    print(p)
  }
  
  dev.off()
  cat(sprintf("✅ Visualizations saved to → %s\n", out_pdf))
}

# ---------------------------------------------------------
# Main Processing: Annotations
# ---------------------------------------------------------
process_annotations <- function(csv_path, img_dir, out_img_dir, out_csv, side, visualize, vis_n) {
  dir.create(out_img_dir, recursive = TRUE, showWarnings = FALSE)
  dir.create(dirname(out_csv), recursive = TRUE, showWarnings = FALSE)
  
  cat(sprintf("\n📦 Loading annotations from %s...\n", csv_path))
  df <- read_csv(csv_path, show_col_types = FALSE)
  
  df <- df %>%
    rename(image = IMAGE_NAME, label = CLASS_NAME) %>%
    mutate(
      label = tolower(trimws(as.character(label))),
      image = trimws(as.character(image))
    )
  
  parsed <- t(sapply(df$GEOMETRY_TEXT, parse_geometry))
  df$TLx <- parsed[, 1]
  df$TLy <- parsed[, 2]
  df$BRx <- parsed[, 3]
  df$BRy <- parsed[, 4]
  
  df <- df %>% filter(!is.na(TLx) & !is.na(image))
  unique_images <- unique(df$image)
  
  cat(sprintf("\n✂️ Splitting %d ANNOTATED images (%s side)...\n", length(unique_images), toupper(side)))
  pb <- progress_bar$new(total = length(unique_images), format = "[:bar] :percent :eta")
  
  new_rows_list <- list()
  
  for (fname in unique_images) {
    pb$tick()
    src <- file.path(img_dir, fname)
    if (!file.exists(src)) next
    
    out_path <- file.path(out_img_dir, fname)
    
    # Crop the image and get the offset coordinate math
    crop_data <- crop_and_save(src, out_path, side)
    mid <- crop_data$mid
    offset <- crop_data$offset
    
    # 🧠 ADJUST COORDINATES
    sub <- df %>% 
      filter(image == fname) %>%
      mutate(
        # Subtract the offset (0 if left, w/2 if right)
        TLx = TLx - offset,
        BRx = BRx - offset
      ) %>%
      # Ensure boxes are actually on the chosen side
      filter(BRx > 0 & TLx < mid) %>%
      mutate(
        # Clip coordinates so they don't bleed off the edge of the new image
        TLx = pmax(TLx, 0),
        BRx = pmin(BRx, mid),
        bw = BRx - TLx,
        bh = BRy - TLy
      ) %>%
      filter(bw > 1 & bh > 1) %>% 
      mutate(
        image = fname,
        TLx = round(TLx, 2),
        TLy = round(TLy, 2),
        BRx = round(BRx, 2),
        BRy = round(BRy, 2)
      ) %>%
      select(image, TLx, TLy, BRx, BRy, label)
    
    if (nrow(sub) > 0) {
      new_rows_list[[fname]] <- sub
    }
  }
  
  new_df <- bind_rows(new_rows_list)
  write_csv(new_df, out_csv, na = "")
  cat(sprintf("\n💾 Saved %d mapped annotations to → %s\n", nrow(new_df), out_csv))
  
  # TRIGGER VISUALIZATION
  if (visualize && nrow(new_df) > 0) {
    vis_pdf_path <- file.path(dirname(out_csv), sprintf("validation_plot_%s.pdf", side))
    visualize_annotations(new_df, out_img_dir, vis_pdf_path, vis_n)
  }
}

# ---------------------------------------------------------
# Main Processing: Zeros
# ---------------------------------------------------------
process_zeros <- function(csv_path, img_dir, out_img_dir, out_csv, side) {
  dir.create(out_img_dir, recursive = TRUE, showWarnings = FALSE)
  dir.create(dirname(out_csv), recursive = TRUE, showWarnings = FALSE)
  
  cat(sprintf("\n📦 Loading zero-annotations from %s...\n", csv_path))
  df <- read_csv(csv_path, show_col_types = FALSE)
  zero_images <- unique(na.omit(df$imagename))
  
  cat(sprintf("\n✂️ Splitting %d ZERO images (%s side)...\n", length(zero_images), toupper(side)))
  pb <- progress_bar$new(total = length(zero_images), format = "[:bar] :percent :eta")
  
  new_rows_list <- list()
  
  for (fname in zero_images) {
    pb$tick()
    src <- file.path(img_dir, fname)
    if (!file.exists(src)) next
    
    out_path <- file.path(out_img_dir, fname)
    crop_and_save(src, out_path, side)
    
    new_rows_list[[fname]] <- data.frame(
      image = fname,
      TLx = NA, TLy = NA, BRx = NA, BRy = NA,
      label = "background",
      stringsAsFactors = FALSE
    )
  }
  
  new_df <- bind_rows(new_rows_list)
  write_csv(new_df, out_csv, na = "")
  cat(sprintf("\n💾 Saved %d background entries to → %s\n", nrow(new_df), out_csv))
}

# ---------------------------------------------------------
# Argument Parsing
# ---------------------------------------------------------
main <- function() {
  option_list <- list(
    make_option("--ann_csv", type = "character"),
    make_option("--ann_img_dir", type = "character"),
    make_option("--out_ann_img_dir", type = "character"),
    make_option("--out_ann_csv", type = "character"),
    
    make_option("--zero_csv", type = "character"),
    make_option("--zero_img_dir", type = "character"),
    make_option("--out_zero_img_dir", type = "character"),
    make_option("--out_zero_csv", type = "character"),
    
    make_option("--side", type = "character", help = "Side to split: 'left' or 'right'"),
    make_option("--visualize", action = "store_true", default = FALSE, help = "Generate a PDF validating the bounding boxes"),
    make_option("--vis_n", type = "integer", default = 5, help = "Number of random images to visualize")
  )
  
  parser <- OptionParser(option_list = option_list)
  opt <- parse_args(parser)
  
  required_args <- c("ann_csv", "ann_img_dir", "out_ann_img_dir", "out_ann_csv", 
                     "zero_csv", "zero_img_dir", "out_zero_img_dir", "out_zero_csv", "side")
  
  if (any(sapply(required_args, function(x) is.null(opt[[x]])))) {
    print_help(parser)
    stop("Missing required arguments.", call. = FALSE)
  }
  
  process_annotations(opt$ann_csv, opt$ann_img_dir, opt$out_ann_img_dir, opt$out_ann_csv, opt$side, opt$visualize, opt$vis_n)
  process_zeros(opt$zero_csv, opt$zero_img_dir, opt$out_zero_img_dir, opt$out_zero_csv, opt$side)
  
  cat("\n✅ All processing complete!\n")
}

main()

File: s/split_script.py
#!/usr/bin/env python3

import argparse
import re
import math
from pathlib import Path
import pandas as pd
from PIL import Image
from tqdm import tqdm

def parse_geometry_to_square(geom_text):
    """
    Parses Habcam GEOMETRY_TEXT.
    Converts the annotated line segment into a square bounding box 
    using the line as the diameter.
    """
    if pd.isna(geom_text):
        return None, None, None, None
    
    nums = re.findall(r"[-+]?\d*\.\d+|\d+", str(geom_text))
    if len(nums) >= 4:
        x1, y1, x2, y2 = map(float, nums[:4])
        
        diameter = math.hypot(x2 - x1, y2 - y1)
        radius = diameter / 2.0
        
        cx = (x1 + x2) / 2.0
        cy = (y1 + y2) / 2.0
        
        return cx - radius, cy - radius, cx + radius, cy + radius
    
    return None, None, None, None

def load_path_mapping(txt_file):
    """Reads the source text manifest into a dictionary of {filename: absolute_path}"""
    print(f"📄 Loading source paths from {txt_file}...")
    mapping = {}
    with open(txt_file, 'r') as f:
        for line in f:
            line = line.strip()
            if line:
                filename = Path(line).name
                mapping[filename] = line
    return mapping

def process_image(src_path, out_path, side, skip_images):
    """
    Reads image dimensions to calculate bounding box offsets.
    If skip_images is False, it also crops and saves the image to disk.
    Returns (None, None, None, None) if the image is corrupted or unreadable.
    """
    try:
        # PIL lazily loads the header here, which is extremely fast for getting dimensions
        with Image.open(src_path) as im:
            w, h = im.size
            mid = w / 2
            
            if side == "left":
                crop_box = (0, 0, mid, h)
                offset = 0
                new_w = mid
            else: # right
                crop_box = (mid, 0, w, h)
                offset = mid
                new_w = w - mid
            
            # Only perform the heavy pixel crop/save if we aren't skipping images
            if not skip_images:
                cropped_im = im.crop(crop_box)
                cropped_im.save(out_path)
            
            return mid, offset, new_w, h
            
    except Exception as e:
        tqdm.write(f"\n[WARN] Skipping unreadable image {Path(src_path).name}: {e}")
        return None, None, None, None

def process_annotations(csv_path, src_txt, out_img_dir, out_csv, side, skip_images):
    """Processes the annotated images, applies bbox math, splits, and remaps while retaining metadata."""
    if not skip_images:
        out_img_dir = Path(out_img_dir)
        out_img_dir.mkdir(parents=True, exist_ok=True)
        
    Path(out_csv).parent.mkdir(parents=True, exist_ok=True)
    path_mapping = load_path_mapping(src_txt)

    print(f"📦 Loading annotations from {csv_path}...")
    df = pd.read_csv(csv_path)

    # Normalize standard columns while keeping all metadata columns intact
    if "IMAGE_NAME" in df.columns:
        df = df.rename(columns={"IMAGE_NAME": "image", "CLASS_NAME": "label"})
    
    df["label"] = df["label"].astype(str).str.strip().str.lower()
    df["image"] = df["image"].astype(str).str.strip()

    # 🔥 EXCLUDE POINT ANNOTATIONS
    print("🧹 Filtering out 'point' annotations...")
    initial_count = len(df)
    df = df[~df["GEOMETRY_TEXT"].astype(str).str.lower().str.contains("point", na=False)].copy()
    print(f"   Dropped {initial_count - len(df)} point annotations.")

    print("🧠 Converting line annotations to square bounding boxes...")
    bounds = df["GEOMETRY_TEXT"].apply(parse_geometry_to_square)
    df["TLx"] = [b[0] for b in bounds]
    df["TLy"] = [b[1] for b in bounds]
    df["BRx"] = [b[2] for b in bounds]
    df["BRy"] = [b[3] for b in bounds]
    
    df = df.dropna(subset=["TLx", "TLy", "BRx", "BRy", "image"])
    unique_images = [img for img in df["image"].unique() if img in path_mapping and Path(path_mapping[img]).exists()]

    processed_dfs = []

    print(f"\n✂️ Processing {len(unique_images)} ANNOTATED images ({side.upper()} side)...")
    for fname in tqdm(unique_images):
        src = path_mapping[fname]
        out_path = out_img_dir / fname if not skip_images else None
        
        crop_result = process_image(src, out_path, side, skip_images)
        if crop_result[0] is None:
            continue
            
        mid, offset, new_w, h = crop_result

        # Isolate the rows for this specific image
        sub = df[df["image"] == fname].copy()
        
        # Shift X coordinates
        sub["TLx"] -= offset
        sub["BRx"] -= offset
        
        # Filter out boxes that are entirely on the wrong side
        sub = sub[sub["BRx"] > 0].copy()
        sub = sub[sub["TLx"] < new_w].copy()
        
        # 🔥 APPLY STRICT 4-WAY CLIPPING
        sub["TLx"] = sub["TLx"].clip(lower=0)
        sub["BRx"] = sub["BRx"].clip(upper=new_w)
        sub["TLy"] = sub["TLy"].clip(lower=0)
        sub["BRy"] = sub["BRy"].clip(upper=h)
        
        # Drop invalid/sliver boxes (must be > 1x1 pixels)
        sub["bw"] = sub["BRx"] - sub["TLx"]
        sub["bh"] = sub["BRy"] - sub["TLy"]
        sub = sub[(sub["bw"] > 1) & (sub["bh"] > 1)]

        # Drop the temporary calculation columns and append the surviving dataframe
        if not sub.empty:
            sub = sub.drop(columns=["bw", "bh"])
            
            # Reorder columns to put image/bbox at the front, followed by all metadata
            cols = ["image", "TLx", "TLy", "BRx", "BRy", "label"]
            other_cols = [c for c in sub.columns if c not in cols]
            sub = sub[cols + other_cols]
            
            processed_dfs.append(sub)

    if processed_dfs:
        new_df = pd.concat(processed_dfs, ignore_index=True)
        # Round the bounding box coordinates
        new_df[["TLx", "TLy", "BRx", "BRy"]] = new_df[["TLx", "TLy", "BRx", "BRy"]].round(2)
        new_df.to_csv(out_csv, index=False)
        print(f"💾 Saved {len(new_df)} mapped annotations to → {out_csv}")
    else:
        print("⚠️ No annotations survived the filtering and splitting process.")

def process_zeros(csv_path, src_txt, out_img_dir, out_csv, side, skip_images):
    """Processes the background images separately while retaining metadata."""
    if not skip_images:
        out_img_dir = Path(out_img_dir)
        out_img_dir.mkdir(parents=True, exist_ok=True)
        
    Path(out_csv).parent.mkdir(parents=True, exist_ok=True)
    path_mapping = load_path_mapping(src_txt)

    print(f"\n📦 Loading zero-annotations from {csv_path}...")
    df = pd.read_csv(csv_path)
    
    img_col = "imagename" if "imagename" in df.columns else "IMAGE_NAME"
    if img_col in df.columns:
        df = df.rename(columns={img_col: "image"})
        
    zero_images = df["image"].dropna().unique()
    valid_zeros = [img for img in zero_images if img in path_mapping and Path(path_mapping[img]).exists()]

    processed_dfs = []

    print(f"\n✂️ Processing {len(valid_zeros)} ZERO images ({side.upper()} side)...")
    for fname in tqdm(valid_zeros):
        src = path_mapping[fname]
        out_path = out_img_dir / fname if not skip_images else None
        
        crop_result = process_image(src, out_path, side, skip_images)
        if crop_result[0] is None:
            continue

        # Keep original metadata for zeros
        sub = df[df["image"] == fname].copy()
        sub["TLx"] = None
        sub["TLy"] = None
        sub["BRx"] = None
        sub["BRy"] = None
        sub["label"] = "background"
        
        # Reorder columns
        cols = ["image", "TLx", "TLy", "BRx", "BRy", "label"]
        other_cols = [c for c in sub.columns if c not in cols]
        sub = sub[cols + other_cols]
        
        processed_dfs.append(sub)

    if processed_dfs:
        new_df = pd.concat(processed_dfs, ignore_index=True)
        new_df.to_csv(out_csv, index=False)
        print(f"💾 Saved {len(new_df)} background entries to → {out_csv}")
    else:
        print("⚠️ No zero-annotation images processed.")

def main():
    parser = argparse.ArgumentParser(description="Split Habcam images and apply Square BBox math directly from manifest.")
    
    # Annotation Arguments
    parser.add_argument("--ann_csv", required=True)
    parser.add_argument("--ann_src_txt", required=True)
    parser.add_argument("--out_ann_img_dir", required=True)
    parser.add_argument("--out_ann_csv", required=True)
    
    # Zero Image Arguments
    parser.add_argument("--zero_csv", required=True)
    parser.add_argument("--zero_src_txt", required=True)
    parser.add_argument("--out_zero_img_dir", required=True)
    parser.add_argument("--out_zero_csv", required=True)
    
    # Global Arguments
    parser.add_argument("--side", choices=["left", "right"], required=True)
    parser.add_argument("--skip_images", action="store_true", help="Skip cropping and saving images; only regenerate CSVs.")

    args = parser.parse_args()

    process_annotations(args.ann_csv, args.ann_src_txt, args.out_ann_img_dir, args.out_ann_csv, args.side, args.skip_images)
    process_zeros(args.zero_csv, args.zero_src_txt, args.out_zero_img_dir, args.out_zero_csv, args.side, args.skip_images)
    
    print("\n✅ All processing complete!")

if __name__ == "__main__":
    main()